### PR TITLE
Fix the bus library to support all the possible address widths

### PR DESCRIPTION
**Pull request by inesmaria08**
### Pull Request Overview

This pull request fixes the `bus` library to support all the possible address types. Before this change, the `bus` library:
 - would error if the address with was larger than 16 bits
 - would possible not be able to store 64 bit addresses as the value of the address was encoded into a `usize` that may be only 32 bits wide (this is the case for all the boards now)
 - the `bus8080` library would accept `usize` address types even though it only supports 16 bit addresses

### Testing Strategy

This pull request was tested on STM32F4 Discovery board and on Raspberry Pi Pico (connected to Pico Explorer Base).

### TODO or Help Wanted

~~This pull request depends on #4098.~~

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

**Comment by alexandruradovici** [comment]
Addresses should always be fixed. I thought about templating it, but I am not sure about the benefit of it. The only thing that changes is the address with and endianness. Do you have any template idea?

File: capsules/extra/src/bus.rs
Comment (lschuermann) [inline capsules/extra/src/bus.rs:152]: Without looking too much at the general interface here, it seems like what we'd really want to have is the bus address be encapsulated within the enum itself, like so:

```
    fn set_addr(&self, addr: BusAddr) -> Result<(), ErrorCode> {
```

and

```
enum BusAddr {
    Bits8([u8; 1]),
    Bits16BE([u8; 2]),
    Bits16LE([u8; 2]),
}
```

@bradjc's suggestion about making it generic also makes sense to me. One implementation of the `BusAddr` trait could still have multiple variants at runtime, but when you know that all your connected peripherals have only one address width, you'd simply plug in the particular type that describes this address.

Something like this could work:

```
struct BusAddr8([u8; 1]);
impl BusAddr for BusAddr8 { ... }
struct BusAddr16BE([u8; 2]);
impl BusAddr for BusAddr16BE { ... }

enum MultiBusAddr {
    Bits8(BusAddr8),
    Bits16BE(BusAddr16BE),
    ...,
}
impl BusAddr for MultiBusAddr { ... }

impl<'a, S: SpiMasterDevice<'a>, A: BusAddr> Bus<'a, A> for SpiMasterBus<'a, S> {
    fn set_addr(&self, addr: impl Into<A>) -> Result<(), ErrorCode> {
```

Pretty sure that the types would need some more massaging to actually work in practice, just some ideas.

Comment (alexandruradovici) [reply capsules/extra/src/bus.rs:152]: I agree with enum, I'm not sure about the templating. Does it really make sense? What about adding a `write_to_slice` function to the enum?

Comment (lschuermann) [reply capsules/extra/src/bus.rs:152]: > I agree with enum, I'm not sure about the templating. Does it really make sense? What about adding a `write_to_slice` function to the enum?

That will still have the `enum BusAddr` be at least 9 bytes in size everywhere, which seems slightly wasteful (although not more wasteful than passing the `enum BusWidth` plus a `u64`).

I think this is a good step in the right direction, and then we can always change it to be a generic later.

**Comment by bradjc** [comment]
Using `u64` on a 32 bit platform always seems like a big hammer to me (particularly for a 16 bit bus). If it is easy to use the appropriate type then I think we should. If not, then it's not a huge deal.

File: capsules/extra/src/bus.rs
Comment (alexandruradovici) [reply capsules/extra/src/bus.rs:152]: Oh I see what you mean now.

Comment (alexandruradovici) [reply capsules/extra/src/bus.rs:152]: I am wondering whether to use a `dyn Trait` instead of a generic? The bus might be used several times on a chip with different address sizes. The generic will duplicate code, while the `dye` trait will perform some small overhead. Not sure which one is better.

**Comment by inesmaria08** [comment]
> Using `u64` on a 32 bit platform always seems like a big hammer to me (particularly for a 16 bit bus). If it is easy to use the appropriate type then I think we should. If not, then it's not a huge deal.

I tested both versions (the previous one that is now in master and this one) and there seems to be no code size impact on ARM architecture. The call stack size might be larger though due to the `u64` argument,.

File: capsules/extra/src/bus.rs
Comment (inesmaria08) [reply capsules/extra/src/bus.rs:152]: Thanks for feedback! I will try to sketch out an implementation in the following days. I see this two possible implementations:

```rust
struct BusAddr<E: Endianess, const N: usize>([u8; N], PhantomData<E>);

trait BusAddrBytes<const N: usize> {
    fn bytes(&self) -> &[u8; N];
}
impl From<u16> for BusAddr<LittleEndian, 2> {
    fn from(value: u16) -> Self {
        Self([value as u8, (value >> 8) as u8], PhantomData)
    }
}
```

or 

```rust
struct BusAddr8([u8; 1]);
struct BusAddr16BE([u8; 2]);
trait BusAddrBytes {
    fn bytes(&self) -> &[u8];
}
impl From<u16> for BusAddr16LE {
    fn from(value: u16) -> Self {
        Self([value as u8, (value >> 8) as u8])
    }
}
...

```
Which one do you think works better?

Comment (lschuermann) [reply capsules/extra/src/bus.rs:152]: I like the second one better, as it explicitly pre-defines the possible address widths. I presume we don't want to support 3-byte addresses?

Also, I'm not sure about `From<u16>` (or `From`-based conversions from integers generally). I more so through of `From` to be useful for if we have a type that reconciles multiple addresses widths at runtime, but maybe that's too complex.

By the way, Rust offers `{to,from}_{le,be}_bytes` functions on integer types which take or produce an array directly. We should prefer those over bitshifts.

Comment (inesmaria08) [reply capsules/extra/src/bus.rs]: I tried the new implementation. I'm waiting for feedback.

File: kernel/src/hil/bus8080.rs
Comment (bradjc) [inline kernel/src/hil/bus8080.rs:19]: I'm not sure how to interpret this. Are the bytes stored in the proper endianess in the associated array? If so, we don't need to know the endianess. If not, why not just use a `u16`?

Comment (lschuermann) [reply kernel/src/hil/bus8080.rs:19]: > Are the bytes stored in the proper endianess in the associated array?

I'm _guessing_ that this is for some imaginary hardware that converts addresses from the target's native endianess to what'll be put on the wire (network byte order?).

Alternatively, there may be buses that have little-endian, and others that have big-endian byte orders. But presumably that would be dictated by the bus itself, and not need to be communicated to / dictated by upper layers?

I'm coming around to this idea. We should just use `u{16,32,64}::{to,from}_be_bytes` to generate the on-wire representation.

Comment (inesmaria08) [reply kernel/src/hil/bus8080.rs:19]: > > Are the bytes stored in the proper endianess in the associated array?
> 
Yes.

> I'm _guessing_ that this is for some imaginary hardware that converts addresses from the target's native endianess to what'll be put on the wire (network byte order?).
> 

No, this is a software problem as the underlying buses are usually serial.

> Alternatively, there may be buses that have little-endian, and others that have big-endian byte orders. But presumably that would be dictated by the bus itself, and not need to be communicated to / dictated by upper layers?

Busses are usually serial (I2C, SPI), meaning they send data byte by byte, without any knowledge about endianness. The receiving device, in out case the screen, is the one dictating the endianness, as it is the one interpreting the data. In the case of Tock, it is the lower-level screen driver that dictates the address endianness. Except for the 8080 bus (which is a parallel one) the endianness is transparent.

> 
> I'm coming around to this idea. We should just use `u{16,32,64}::{to,from}_be_bytes` to generate the on-wire representation.

We do use these functions when we generate the addresses in the implementation of the `From` trait.

Comment (inesmaria08) [reply kernel/src/hil/bus8080.rs:19]: > I'm not sure how to interpret this. Are the bytes stored in the proper endianess in the associated array? If so, we don't need to know the endianess. If not, why not just use a `u16`?

The bytes are stored in the proper endianness, ready to be sent on a serial bus. Even if the bytes are in the proper order, we still need to know the endianness due to buses, like the 8080, that are serial.

We use an array instead of an `u16` as the conversion is done in the `From` trait. We could use an `u16` and defer the conversion to the `BusAddr::bytes` function, but this implies creating a new array every time this function is called.

Comment (alexandruradovici) [reply kernel/src/hil/bus8080.rs:19]: If we use a serial bus, there is no problem. For buses like 8080, which is parallel, the endianness matters.

Comment (bradjc) [reply kernel/src/hil/bus8080.rs:19]: Ok I think I finally understand this. In my opinion it makes more sense to just store the address as a `BusAddr16(u16)` and then convert from/to whatever byte order the bus is using. But if it is cleaner/simpler to just store the bytes as they came in and what endianness they were when they were stored that works too.

But a comment explaining this would be helpful.

Comment (lschuermann) [reply kernel/src/hil/bus8080.rs:19]: I have a preference for `BusAddr16(u16)`.

Comment (bradjc) [reply kernel/src/hil/bus8080.rs:19]: What is target endianness? The address is just the address (stored as a u16). The user of the address just puts the MS byte (`addr >> 8`) wherever it needs and the LS byte (`addr & 0xFF`) wherever it needs.

Comment (lschuermann) [reply kernel/src/hil/bus8080.rs:19]: I guess it might help to have some concrete examples to work through here.

@bradjc Your suggestion makes sense when endianness is always determined by the bus itself.

When two peripherals on a bus can have different edianness expectations for their addresses, then encoding it in the type makes sense.

It never makes sense to have it both in the enum variant `{BE,LE}`, _and_ implicitly in an array `[_; u8]`.

@inesmaria08 @alexandruradovici Can you provide some concrete examples here?

Comment (bradjc) [reply kernel/src/hil/bus8080.rs:19]: > When two peripherals on a bus can have different edianness expectations for their addresses, then encoding it in the type makes sense.

That seems like a very odd design. But ok. I would probably argue that half of the addresses are wrong, but, systems can be complicated.

Comment (inesmaria08) [reply kernel/src/hil/bus8080.rs:19]: Based on what I've read, it seems we tend keep the changes as they are. I'm uncertain about how to proceed from here.

Comment (bradjc) [reply kernel/src/hil/bus8080.rs:19]: I think this just needs documentation. The enum variant encodes what endianness the device uses on the bus. The bytes are then stored somehow, and that somehow should be documented. Is byte index 0 always sent first? Or is byte index 0 always the least significant byte? Or the most significant byte?

**Comment by alexandruradovici** [comment]
Is there anything else that we need for this PR?

**Comment by alevy** [comment]
We need approvals. @alexandruradovici maybe can start with you?

**Comment by lschuermann** [comment]
@alevy I don't think this has addressed the concerns from @bradjc and myself in this thread yet: https://github.com/tock/tock/pull/4099#discussion_r1734677902

**Comment by alevy** [comment]
See https://github.com/tock/tock/pull/4099#discussion_r1734677902 for current feedback state

File: capsules/extra/src/bus.rs
Comment (lschuermann) [inline capsules/extra/src/bus.rs:58]: Same here, no longer using arrays.

**Review by lschuermann — CHANGES_REQUESTED** [review]
I think this is good to go modulo the mismatch between documentation and implementation. I think you should copy (and adjust) that large document over `DataWidth` into the HIL and explain why the various widths & endianness selectors are required.

File: capsules/extra/src/bus.rs
Comment (inesmaria08) [reply capsules/extra/src/bus.rs]: > I think you should copy (and adjust) that large document over DataWidth into the HIL and explain why the various widths & endianness selectors are required.

Since we no longer store the address in an array based on endianness, the comment explaining it has been revised to fit the new implementation and is now much shorter. Do you think we should still include it in the HIL, or is it fine as it is?

File: kernel/src/hil/bus8080.rs
Comment (alexandruradovici) [inline kernel/src/hil/bus8080.rs:19]: I am not sure that we need this comment.

File: capsules/extra/src/bus.rs
Comment (alexandruradovici) [inline capsules/extra/src/bus.rs:59]: Please add at `///` comment to each structure, similar tu:

```rust
/// 64 bit Little Endian Bus Address
```

File: capsules/extra/src/bus.rs
Comment (alexandruradovici) [inline capsules/extra/src/bus.rs:125]: What about having a default implementation here
```suggestion
    fn len(&self) -> usize {
      Self::DATA_WIDTH.width_in_bytes()
    }
```

or delete it completely and use `<addr as BusAddr>::DATA_WIDTH.width_in_bytes()`?

File: capsules/extra/src/bus.rs
Comment (alexandruradovici) [inline capsules/extra/src/bus.rs:132]: This would be replaced by the trait's default implementation.

File: capsules/extra/src/bus.rs
Comment (alexandruradovici) [inline capsules/extra/src/bus.rs:449]: You should put the buffer back here.

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

**Review by alexandruradovici — APPROVED** [review]
LGTM

**Comment by inesmaria08** [comment]
I would like to move this forward. Is there anything you’d like me to adjust, or is it ready to go?
@lschuermann @bradjc

File: kernel/src/hil/bus8080.rs
Comment (alexandruradovici) [reply kernel/src/hil/bus8080.rs:22]: @inesmaria08 can you please add a `///` comment for each of these items?

File: kernel/src/hil/bus8080.rs
Comment (bradjc) [inline kernel/src/hil/bus8080.rs:25]: Can we say something like "Enum to represent the address of a bus-attached device. For addresses larger than a single byte the enum variant captures the endianess used by the device on the bus. The address is stored in the host endianess in the `u16` and must be converted to the correct endianess before using the address on the bus."

**Comment by inesmaria08** [comment]
@bradjc @lschuermann Is there anything else that needs to be added, or is this ready to be merged?

File: capsules/extra/src/bus.rs
Comment (alexandruradovici) [inline capsules/extra/src/bus.rs:296]: Can we move these two lines within the `if`?

**Review by bradjc — DISMISSED** [review]

**Review by bradjc — APPROVED** [review]

**Review by alexandruradovici — APPROVED** [review]

**Review by lschuermann — APPROVED** [review]

**Pull request merged**
